### PR TITLE
crossdock/oneway: Enable oneway raw

### PR DESCRIPTION
**Pull request by vivekrsharma**
This integrates the cross language tests with Go.
Currently uses raw encoding.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -21,8 +21,8 @@ coverage:
 	./gradlew jacocoRootReport
 
 dockclean:
-	docker-compose kill
-	docker-compose rm --force
+	docker-compose kill java
+	docker-compose rm -f java
```
Comment (abhinav) [inline Makefile]: IMO you should change this back. `clean` should mean clean everything, including dependencies.

Feel free to drop the dependency on clean in the `crossdock` target and/or clean-only Java there explicitly.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/OnewayHandler.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import io.yarpc.core.transport.Headers;
+import io.yarpc.core.transport.TransportBody;
+import io.yarpc.core.transport.TransportRequest;
+import io.yarpc.crossdock.server.OnewaySubject;
+import io.yarpc.encoding.raw.RawEncoding;
+import io.yarpc.transport.http.HttpOutbound;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/** OnewayHandler is a mock for oneway. */
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/OnewayHandler.java]: it's not a mock

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/OnewayHandler.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+public class OnewayHandler implements OnewayRequestHandler<byte[]> {
+  public static final String CALL_BACK_ADDR = "callBackAddr";
+  private static final String ONEWWAY_SERVER = "oneway-server";
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/OnewayHandler.java]: typo: `ONEWAY_SERVER`

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/OnewayHandler.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  private static final String ONEWAY_CLIENT = "oneway-client";
+  private static final String CALLBACK = "call-back";
+
+  @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION")
+  @Override
+  public CompletionStage<Void> handleOneway(Context ctx, byte[] body) {
+    callHome(ctx, ctx.getHeaders().get(CALL_BACK_ADDR), body);
+    return CompletableFuture.completedFuture(null);
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/OnewayHandler.java]: You should make callHome return the `Single<Ack>` you get by calling `callOneway` and build a CompletableFuture from that instead of returning this one always. That way if the outbound call fails immediately, this doesn't count as a success.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import io.yarpc.transport.http.HttpTransport;
+
+public class OnewaySubject implements Subject {
+  private static final int HTTP_PORT = 8084;
+  public static final String SERVICE_NAME = "oneway-server";
+  private Dispatcher serverDispatcher;
+  public static final HttpTransport httpTransport = new HttpTransport();
+
+  @Override
+  /** Starts the transport and server dispatcher for Yarpc. */
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java]: docs go above the annotations

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import io.yarpc.crossdock.behavior.oneway.OnewayHandler;
+import io.yarpc.encoding.raw.RawEncoding;
+import io.yarpc.transport.http.HttpInbound;
+import io.yarpc.transport.http.HttpTransport;
+
+public class OnewaySubject implements Subject {
+  private static final int HTTP_PORT = 8084;
+  public static final String SERVICE_NAME = "oneway-server";
+  private Dispatcher serverDispatcher;
+  public static final HttpTransport httpTransport = new HttpTransport();
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java]: why is the transport static but not the dispatcher? IMO none of this should be static.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  @Override
+  /** Starts the transport and server dispatcher for Yarpc. */
+  public void start() {
+    HttpInbound inbound = httpTransport.inboundBuilder().port(HTTP_PORT).build();
+    serverDispatcher = Dispatcher.builder().name(SERVICE_NAME).inbound(inbound).build();
+    serverDispatcher.register(RawEncoding.newOnewayProcedure("echo/raw", new OnewayHandler()));
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java]: Preferably, instantiate everything in the constructor and start it in start.

So this start ends up being just `serverDispatcher.start()`.

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java [hunk 8]
```diff
@@ -176,6 +178,11 @@ protected void channelRead0(ChannelHandlerContext ctx, TransportRequest request)
           handlerSpec.getUnary().handleUnary(request).subscribe(new UnaryResponseHandler(ctx));
           break;
         case Oneway:
+          TransportResponse response =
+              DefaultTransportResponse.builder()
+                  .body(TransportBody.fromByteArray(new byte[0]))
+                  .build();
```
Comment (abhinav) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java]: This seems like a change separate from the rest of this PR. Please pull this into a different PR.

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java [hunk 9]
```diff
@@ -217,16 +224,12 @@ public void onSubscribe(Disposable disposable) {
 
     @Override
     public void onComplete() {
-      // TODO close channel as soon as request is done reading rather than after handling the
-      // response
-      ctx.close();
+      // nothing to do
```
Comment (abhinav) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java]: If we're not doing anything for all other cases, it may be preferable to use [doOnError].

  [doOnError]: http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Completable.html#doOnError(io.reactivex.functions.Consumer)

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java [hunk 10]
```diff
@@ -217,16 +224,12 @@ public void onSubscribe(Disposable disposable) {
...
     }
 
     @Override
     public void onError(Throwable cause) {
-      // TODO send back 500 error instead
-      // TODO log failure
-      ctx.close();
+      // TODO: handle errors
```
Comment (abhinav) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java]: We don't want to fire exception caught after we've closed the channel (which is possible if we finished writing to it before we finished processing the oneway request). We should log the error instead.

File: Makefile (hunk 1 above)
Comment (vivekrsharma) [inline Makefile]: Fixed.

File: yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java (hunk 9 above)
Comment (vivekrsharma) [inline yarpc-transport-http/src/main/java/io/yarpc/transport/http/HttpInbound.java]: Pulled out the HttpInbound.java in another PR

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/RawOnewayHandler.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  public CompletionStage<Void> handleOneway(Context ctx, byte[] body) {
+    return callHome(ctx, ctx.getHeaders().get(CALL_BACK_ADDR), body)
+        .to(CompletionStages::fromSingle)
+        .thenApply(ack -> null);
+  }
+
+  private Single<Ack> callHome(Context ctx, String callBackAddr, byte[] body) {
+    HttpTransport httpTransport = new HttpTransport();
+    HttpOutbound outbound = httpTransport.outboundBuilder().url("http://" + callBackAddr).build();
+    outbound.start();
```
Comment (bombela) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/RawOnewayHandler.java]: you never stop this outbound

File: docker-compose.yml [hunk 12]
```diff
@@ -26,11 +28,12 @@ services:
             - AXIS_HTTPSERVER=java  # TODO: go
             - AXIS_CLIENT_ONEWAY=go
```
Comment (vivekrsharma) [inline docker-compose.yml]: we don't have client for java yet.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/RawOnewayHandler.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+              .caller(ONEWAY_SERVER)
+              .service(ONEWAY_CLIENT)
+              .procedure(CALLBACK)
+              .encoding(RawEncoding.ENCODING)
+              .body(TransportBody.fromByteArray(body))
+              .build();
+
+      return outbound.callOneway(transportRequest);
+    } finally {
+      outbound.stop();
```
Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/RawOnewayHandler.java]: Yes, that's async java :) Done.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/RawOnewayHandler.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        DefaultTransportRequest.builder()
+            .headers(new Headers(ctx.getHeaders()))
+            .caller(ONEWAY_SERVER)
+            .service(ONEWAY_CLIENT)
+            .procedure(CALLBACK)
+            .encoding(RawEncoding.ENCODING)
+            .body(TransportBody.fromByteArray(body))
+            .build();
+
+    return outbound.callOneway(transportRequest).doFinally(() -> outbound.stop());
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/behavior/oneway/RawOnewayHandler.java]: nit: I think you can do `outbound::stop` here.

    doFinally(outbound::stop)

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  /** Starts the transport and server dispatcher for Yarpc. */
+  @Override
+  public void start() {
+    serverDispatcher.start();
+  }
+
+  /** Stops the server dispatcher. */
+  @Override
+  public void stop() {
+    if (serverDispatcher == null) {
```
Comment (abhinav) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/server/OnewaySubject.java]: the server is never null
